CFE-254: Adds Node observability Operator section (#47075)
openshift/openshift-docs: scalability_and_performance/understanding-node-observability-operator.adoc

:_content-type: ASSEMBLY
[id="using-node-observability-operator"]
= Understanding the Node Observability Operator
include::_attributes/common-attributes.adoc[]
:context: node-observability-operator

toc::[]

:FeatureName: The Node Observability Operator
include::snippets/technology-preview.adoc[leveloffset=+0]

> The Node Observability Operator collects and stores the CRI-O and Kubelet profiling data of worker nodes. You can use the profiling data to analyse the CRI-O and Kublet performance trends and debug the performance related issues.

Author: fixed.

+
[source,yaml]
----
    apiVersion: nodeobservability.olm.openshift.io/v1alpha1
    kind: NodeObservability
    metadata:
      name: cluster <1>
    spec:
      labels:
        node-role.kubernetes.io/worker: ""
      type: crio-kubelet
----
<1> You must specify the name as `cluster` because there should be only one `NodeObservability` CR per cluster.

. Run the `NodeObservability` CR:
+
[source,terminal]
----
oc apply -f nodeobservability.yaml
----

+
.Example output
[source,terminal]
----
nodeobservability.olm.openshift.io/cluster created
----

. Review the status of the `NodeObservability` CR:
+
[source,terminal]
----
$ oc get nob/cluster -o yaml | yq '.status'
----

+
.Example output
[source,terminal]
----
conditions:
  conditions:
  - lastTransitionTime: "2022-07-05T07:33:54Z"
    message: 'DaemonSet node-observability-ds ready: true NodeObservabilityMachineConfig
      ready: true'
    reason: Ready
    status: "True"
    type: Ready
    count: 1
    lastUpdated: "2022-07-05T07:33:55Z"
----

+
`NodeObservability` CR run is completed when the reason `Ready` and status `True`.

== Running profiling queries

> Profiling query is a blocking operation that fetches CRI-O and Kubelet profiling for a duration of 30 seconds. To request a profiling query, you have to create a `NodeObservabilityRun` resource.

Reviewer: > for a duration of 30 seconds
Cannot find this timeout anywhere. You meant the duration of the block which the query puts on or the duration of the profiling data?

Author: @alebedev87 I meant duration of the block which the query puts. If this is causing confusion, please suggest a rewrite.

Reviewer: Ah, ok I see [it in the code](https://github.com/openshift/node-observability-agent/blob/main/pkg/handlers/handlers.go#L223). This is a timeout which gives up the query. It's set to 35 seconds actually. 
Well we can remove the mention of the seconds at all, or we can say that it may block up to `35` seconds max.

Reviewer 2: For background, the 30s are probably coming from the duration of the pprof command. 
it's true that 30s is the time it takes when all goes right, and maxed at 35s in case of problem

Reviewer: Yes, true, 30 seconds is [the default probing period for the pprof handler](https://pkg.go.dev/net/http/pprof#Profile) integrated into kubelet.
So, then if I read the phrase right - 30s is the duration of the probe, not the blocking period.

Reviewer: Sorry, about the confusion. The phrase is fine as it was from the beginning.

----

. Switch back to the `node-observability-operator` namespace:
+
[source,terminal]
----
$ oc project node-observability-operator
----

. Create a CR file named `nodeobservability.yaml` that contains the following text:
+
[source,yaml]
----
    apiVersion: nodeobservability.olm.openshift.io/v1alpha1
    kind: NodeObservability
    metadata:
      name: cluster <1>
    spec:
      labels:
        node-role.kubernetes.io/worker: ""
      type: crio-kubelet
----
<1> You must specify the name as `cluster` because there should be only one `NodeObservability` CR per cluster.

. Run the `NodeObservability` CR:
+
[source,terminal]
----
oc apply -f nodeobservability.yaml
----

+
.Example output
[source,terminal]
----
nodeobservability.olm.openshift.io/cluster created
----

. Review the status of the `NodeObservability` CR:
+
[source,terminal]
----
$ oc get nob/cluster -o yaml | yq '.status'
----

+
.Example output
[source,terminal]
----
conditions:
  conditions:
  - lastTransitionTime: "2022-07-05T07:33:54Z"
    message: 'DaemonSet node-observability-ds ready: true NodeObservabilityMachineConfig
      ready: true'
    reason: Ready
    status: "True"
    type: Ready
    count: 1

>     lastUpdated: "2022-07-05T07:33:55Z"

Author: fixed.

----

=== Installing the Node Observability Operator using the web console
You can install the Node Observability Operator from the {product-title} web console.

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.
* You have access to the {product-title} web console.

.Procedure

. Log in to the OpenShift Container Platform web console.
. In the Administrator's navigation panel, expand *Operators* → *OperatorHub*.
. In the *All items* field, enter *Node Observability Operator* and select the Node Observability Operator tile.
. Click *Install*.
. On the *Install Operator* page:
.. *Update channel*: Select *alpha*.
.. *Installation mode*: Select *A specific namespace on the cluster*.
.. *Installed Namespace*: Choose *node-observability-operator* from the list.
.. *Update approval*: Select *Automatic*.
.. Click *Install*.

.Verification
. In the Administrator's navigation panel, expand *Operators* → *Installed Operators*.
. Verify that the Node Observability Operator is listed in the Operators list.

== Creating the Node Observability custom resource
Before you run profiling queries, you must create a `NodeObservability` custom resource (CR).

> [IMPORTANT]

Author: @alebedev87 @sherine-k  Please review and let me know if IMPORTANT block make sense.

Reviewer: Yes, it makes sense, thanks for adding this!

====

> Creation of `NodeObservability` CR will reboot all the nodes. It might take 10 or more minutes to complete.

Reviewer: ```suggestion
Creation of `NodeObservability` CR will reboot all the worker nodes. It might take 10 or more minutes to complete.
```

Author: fixed.

====

> When you apply the the `NodeObservability` CR, it creates the necessary machine config and machine config pool CRs to enable the CRI-O profiling on selected worker nodes.

Author: fixed.

[NOTE]
====
Kubelet profiling is enabled by default.
====

> The CRIO unix socket of the node is mounted on the agent pod, which allows the agent to communicate with CRIO to run the pprof request. Similiarly, the `kubelet-serving-ca` certificate chain is mounted on the agent pod, which allows secure communication between the agent and node's kubelet endpoint.

Author: @alebedev87 @sherine-k Please review this content.

:_content-type: ASSEMBLY
[id="using-node-observability-operator"]
= Understanding the Node Observability Operator
include::_attributes/common-attributes.adoc[]
:context: node-observability-operator

toc::[]

:FeatureName: The Node Observability Operator
include::snippets/technology-preview.adoc[leveloffset=+0]

The Node Observability Operator collects and stores the CRI-O and Kubelet profiling data of worker nodes. You can use the profiling data to analyse the CRI-O and Kublet performance trends and debug the performance related issues.

== High level workflow of the Node Observability Operator

> When you install a Node Observability Operator in {product-title} cluster, it creates a DaemonSet to deploy a Node Observability agent on each worker node. To request a profiling query, you have to create a `NodeObservabilityRun` resource, which requests the deployed Node Observability agent to trigger the CRI-O and Kubelet profiling. After the profiling is completed, Node Observability agent stores the profiling data inside the container file system `/run` directory, which is available for query over the time.

Author: Fixed.

== Installing the Node Observability Operator
The Node Observability Operator is not installed in {product-title} by default. You can install the Node Observability Operator by using the {product-title} CLI or the web console.

=== Installing the Node Observability Operator using the CLI

.Prerequisites

* You have installed the OpenShift CLI (oc).
* You have access to the cluster with `cluster-admin` privileges.

.Procedure

. Confirm that the Node Observability Operator is available:
+
[source,terminal]
----

> $ oc get packagemanifests -n openshift-marketplace node-observability-operator

Reviewer: Example can be added:
```
NAME               CATALOG                AGE
node-observability-operator   Red Hat Operators      9h
```

Author: fixed.

----

> . Create the `node-observability` Operator namespace:

Author: fixed.

+
[source,terminal]
----
$ oc new-project node-observability-operator
----

. Create an `OperatorGroup` object YAML file:
+
[source,yaml]
----
cat <<EOF | oc apply -f -
apiVersion: operators.coreos.com/v1
kind: OperatorGroup
metadata:
  name: node-observability-operator
  namespace: node-observability-operator
spec:
  targetNamespaces:
  - node-observability-operator
EOF
----

. Create a `Subscription` object YAML file to subscribe a namespace to an Operator:
+
[source,yaml]
----
cat <<EOF | oc apply -f -
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: node-observability-operator
  namespace: node-observability-operator
spec:
  channel: alpha
  name: node-observability-operator

>   source: node-observability-operator

Reviewer: ```suggestion
  source: redhat-operators
```

  sourceNamespace: openshift-marketplace
EOF
----

.Verification

. View the install plan name:
+
[source,terminal]
----
$ oc -n node-observability-operator get sub node-observability-operator -o yaml | yq '.status.installplan.name'
----

+
.Example output
[source,terminal]
----
install-dt54w
----

. Verify the install plan status:
+
[source,terminal]
----

> $ oc -n node-observability-operator get ip <install_plan_name> -o yaml | yq .status.phase'

Author: Fixed.

----
+

> `<install_plan_name>` is the install plan name that you obtained from the output of the previous command.

Reviewer: It'll be different for each user, if this obliges us to put a placeholder - I'm fine with this.

+
.Example output
[source,terminal]
----
COMPLETE
----

. Verify that the Node Observability Operator is up and running:
+
[source,terminal]
----
$ oc get deploy -n node-observability-operator
----

+
.Example output
[source,terminal]
----
NAME                                            READY   UP-TO-DATE  AVAILABLE   AGE
node-observability-operator-controller-manager  1/1     1           1           40h
----

=== Installing the Node Observability Operator using the web console
You can install the Node Observability Operator from the {product-title} web console.

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.
* You have access to the {product-title} web console.

.Procedure

. Log in to the OpenShift Container Platform web console.
. In the Administrator's navigation panel, expand *Operators* → *OperatorHub*.

> . In the *All items* field, enter `node-observability-operator` and select the Node Observability Operator tile.

Author: fixed.

. Click *Install*.
. On the *Install Operator* page:
.. *Update channel*: Select *alpha*.
.. *Installation mode*: Select *A specific namespace on the cluster*.
.. *Installed Namespace*: Choose *node-observability-operator* from the list.
.. *Update approval*: Select *Automatic*.
.. Click *Install*.

.Verification
. In the Administrator's navigation panel, expand *Operators* → *Installed Operators*.
. Verify that the Node Observability Operator is listed in the Operators list.

== Creating the Node Observability custom resource
Before you run profiling queries, you must create a `NodeObservability` custom resource (CR).

> When you run the the `NodeObservability` CR, it creates the necessary machine config and machine config pool CRs to enable the CRI-O profiling on selected worker nodes.

Author: fixed.

[NOTE]
====
Kubelet profiling is enabled by default.
====

.Prerequisites
* The Node Observability Operator has been installed.
* You have access to the cluster with `cluster-admin` privileges.

.Procedure

. Log in to the {product-title} CLI as a user with the `cluster-admin` role:
+
[source,terminal]
----
$ oc login -u kubeadmin https://<HOSTNAME>:6443
----

. Create a new project named `node-observability-operator`:
+
[source,terminal]
----
$ oc project node-observability-operator

> ----

Author: ok. I will remove this step. Thanks.

. Create a CR file named `nodeobservability.yaml` that contains the following text:
+
[source,yaml]
----
    apiVersion: nodeobservability.olm.openshift.io/v1alpha1
    kind: NodeObservability
    metadata:
      name: cluster <1>
    spec:

>       labels:

Reviewer: ```suggestion
      labels:
        node-role.kubernetes.io/worker: ""
```
Better to specify a concrete label to so that the agent gets deployed on the worker nodes not by the virtue of not having the right tolerations to be on the master nodes.

Author: fixed.

      type: crio-kubelet
----

> <1> You must specify the name as `cluster` because it is applied cluster-wide.

Author: @alebedev87 How about this rewrite ? wdyt ?

<1> You must specify the name as `cluster` because there can be only one `NodeObservability` CR per cluster.

. Run the `NodeObservability` CR:
+
[source,terminal]
----
oc apply -f nodeobservability.yaml
----

+
.Example output
[source,terminal]
----
nodeobservability.olm.openshift.io/cluster created
----

. Review the status of the `NodeObservability` CR:
+
[source,terminal]
----

> $ oc get nob/cluster -o yaml

Author: Fixed.

    finalizers:
    - NodeObservability
    generation: 1
    name: cluster
    resourceVersion: "131690"
    uid: 897815d9-929d-48d9-9f1a-53f1390b456c
  spec:
    type: crio-kubelet
  status:
    conditions:
      conditions:
      - lastTransitionTime: "2022-07-05T07:33:54Z"
        message: 'DaemonSet node-observability-ds ready: true NodeObservabilityMachineConfig
          ready: true'
        reason: Ready
        status: "True"
        type: Ready
    count: 1
    lastUpdated: "2022-07-05T07:33:55Z"
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
----

+
`NodeObservability` CR run is completed when the reason `Ready` and status `True`.

== Running profiling queries

Profiling query is a blocking operation that fetches CRI-O and Kubelet profiling for a duration of 30 seconds. To request a profiling query, you have to create a `NodeObservabilityRun` resource.

[IMPORTANT]
====
You can request only one profiling query at any point of time.
====

.Prerequisites
* The Node Observability Operator has been installed.
* The `NodeObservability` custom resource (CR) has been created.
* You have access to the cluster with `cluster-admin` privileges.

.Procedure

> . Create a `NodeObservabilityRun` resource file named `nodeobservabilityrun.yaml` that contains the following text:

Author: @alebedev87 Did you mean 

> . Create the NodeObservabilityRun CR namespace:
> 
> $ oc project node-observability-operator

> . Create a `NodeObservabilityRun` resource file named `nodeobservabilityrun.yaml` that contains the following text:

Author: @alebedev87 Sherine told me to remove `$ oc project node-observability-operator` step. She thinks there's no need to recreate project as the user would have done it at the previous step while installing the operator.

@sherine-k Could you please pitch in and share your thoughts.

> . Create a `NodeObservabilityRun` resource file named `nodeobservabilityrun.yaml` that contains the following text:

Reviewer: `oc project` command doesn't create a project, it switches to the the specified project. You have to switch to it because all the commands for `NodeObservabilityRun` have to be done in the project or changed to add the namespace.

> . Create a `NodeObservabilityRun` resource file named `nodeobservabilityrun.yaml` that contains the following text:

Reviewer 2: Hi I agree we should keep `oc project node-observability-operator`.
This command switches to the existing namespace `node-observability-operator`.
My comment was more to say that we don't have to create the namespace again. I must have not read properly the command. So sorry about that

> . Create a `NodeObservabilityRun` resource file named `nodeobservabilityrun.yaml` that contains the following text:

Author: Fixed and pushed the changes.

> . Create a `NodeObservabilityRun` resource file named `nodeobservabilityrun.yaml` that contains the following text:

Author: @alebedev87 Do you think example output is needed for this command ? If yes, please share.

> . Create a `NodeObservabilityRun` resource file named `nodeobservabilityrun.yaml` that contains the following text:

Reviewer: No, just `oc project node-observability-operator` command is enough.

      {"apiVersion":"nodeobservability.olm.openshift.io/v1alpha1","kind":"NodeObservabilityRun","metadata":{"annotations":{},"name":"nodeobservabilityrun-sample","namespace":"node-observability-operator"},"spec":{"nodeObservabilityRef":{"name":"cluster"}}}
  creationTimestamp: "2022-07-07T14:57:34Z"
  generation: 1
  name: nodeobservabilityrun-sample
  namespace: node-observability-operator
  ownerReferences:
  - apiVersion: nodeobservability.olm.openshift.io/v1alpha1
    kind: NodeObservability
    name: cluster
    uid: d544a832-2100-4574-8eee-aec70544c5ed
  resourceVersion: "297555"
  uid: 7b77c188-2e7f-4796-9c3d-170a0f873152
spec:
  nodeObservabilityRef:
    name: cluster
status:
  agents:
  - ip: 10.217.0.68
    name: node-observability-ds-wvrxc
    port: 8443
  conditions:
    conditions:
    - lastTransitionTime: "2022-07-07T14:57:34Z"
      message: Ready to start profiling
      reason: Ready
      status: "True"
      type: Ready
    - lastTransitionTime: "2022-07-07T14:58:10Z"
      message: Profiling query done
      reason: Finished
      status: "True"
      type: Finished
  finishedTimestamp: "2022-07-07T14:58:10Z"
  startTimestamp: "2022-07-07T14:57:35Z"
----

+

> Profiling query is complete when the status `True` and type `Finished`.

Author: @alebedev87 Did you mean
Profiling query is complete when the reason `Finished` and  status `True`?

> Profiling query is complete when the status `True` and type `Finished`.

Reviewer: Yes, it's complete when the status for `Finished` reason is `True`.

> . Run the following bash script to store the profiling data to `/run/node-observability` path:

Reviewer: ```suggestion
. Run the following bash script to retrieve the profiling data from container's `/run/node-observability` path:
```

Author: fixed.

+
[source,bash]
----
for i in `oc get nodeobservabilityrun.nodeobservability.olm.openshift.io/nodeobservabilityrun -o yaml | yq .status.agents[].name | cut -d\" -f2`
  do
  echo $i
  list=`oc exec $i -c node-observability-agent -- bash -c "ls /run/node-observability/*.pprof"`
  for j in $list
    do
    k=`echo $j|cut -d\/ -f3`
    mkdir -p /tmp/$i
    echo copying $k to /tmp/$i/$k
    kubectl exec $i -c node-observability-agent -- cat $j > /tmp/$i/$k
  done

> done

Reviewer: Smaller and stricter bash version of this script:
```
for a in $(oc get nodeobservabilityrun nodeobservabilityrun -o yaml | yq .status.agents[].name); do
  echo "agent ${a}"
  mkdir -p "/tmp/${a}"
  for p in $(oc exec "${a}" -c node-observability-agent -- bash -c "ls /run/node-observability/*.pprof"); do
    f="$(basename ${p})"
    echo "copying ${f} to /tmp/${a}/${f}"
    oc exec "${a}" -c node-observability-agent -- cat "${p}" > "/tmp/${a}/${f}"
  done
done
```

Reviewer: Manually tested:
```
$ ./script.sh
agent node-observability-ds-zp24r
copying crio-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof to /tmp/node-observability-ds-zp24r/crio-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof
copying kubelet-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof to /tmp/node-observability-ds-zp24r/kubelet-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof
agent node-observability-ds-szdf4
copying crio-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof to /tmp/node-observability-ds-szdf4/crio-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof
copying kubelet-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof to /tmp/node-observability-ds-szdf4/kubelet-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof
agent node-observability-ds-wct6p
copying crio-865dc234-2b0d-444e-bd18-44333355ed77.pprof to /tmp/node-observability-ds-wct6p/crio-865dc234-2b0d-444e-bd18-44333355ed77.pprof
copying kubelet-865dc234-2b0d-444e-bd18-44333355ed77.pprof to /tmp/node-observability-ds-wct6p/kubelet-865dc234-2b0d-444e-bd18-44333355ed77.pprof

$ ls -ltr /tmp/node-observability-ds-*
/tmp/node-observability-ds-zp24r:
total 32
-rw-rw-r--. 1 alebedev alebedev  7060 Jul 11 12:11 crio-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof
-rw-rw-r--. 1 alebedev alebedev 22511 Jul 11 12:11 kubelet-0d2edeed-9c9f-4f75-9cc7-0cb5f79ac0d4.pprof

/tmp/node-observability-ds-szdf4:
total 32
-rw-rw-r--. 1 alebedev alebedev  6635 Jul 11 12:11 crio-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof
-rw-rw-r--. 1 alebedev alebedev 23364 Jul 11 12:11 kubelet-65e80d14-3961-49fa-9717-5bb6f55367a6.pprof

/tmp/node-observability-ds-wct6p:
total 28
-rw-rw-r--. 1 alebedev alebedev  3345 Jul 11 12:12 crio-865dc234-2b0d-444e-bd18-44333355ed77.pprof
-rw-rw-r--. 1 alebedev alebedev 21965 Jul 11 12:12 kubelet-865dc234-2b0d-444e-bd18-44333355ed77.pprof
```

Author: Fixed.

== Key features
* Trigger a CRIO-O and kublet profiling.
* You can deploy Node Observability Agents on selected nodes.
* Users can access the profiling data files for analysis.

== Installing the Node Observability Operator using the web console
The Node Observability Operator is not installed in {product-title} by default. You can install the Node Observability Operator by using the {product-title} web console.

.Prerequisites

* You have access to the cluster with `cluster-admin` privileges.
* You have access to the {product-title} web console.

.Procedure

. Log in to the OpenShift Container Platform web console.
. Navigate to *Operators* → *OperatorHub*.
. Enter `node-observability-operator` into the filter box to locate the Node Observability Operator.
. Click *Install*.
. On the *Install Operator* page:
.. *Update channel*: Select *alpha*.
.. *Installation mode*: Select *A specific namespace on the cluster*.
.. *Installed Namespace*: Choose *node-observability-operator* from the list.
.. *Update approval*: Select *Automatic*.
.. Click *Install*.

== Deploying the Node Observability Operator

> Before you can deploy the Node Observability Operator, you should create a `NodeObservability` custom resource (CR).

Reviewer 2: Hi @darshan-nagaraj 
We don't really need `NodeObservability` custom resource in order to deploy the operator. 
The operator will be deployed by the above procedure, and will handle `NodeObservability` CRs once created.
What you can say however, is that `NodeObservability` is needed in order to create a `NodeObservabilityRun`, which is the profiling request.

Author: Fixed.

.Prerequisites
* The Node Observability Operator has been installed.
* You have access to the cluster as a user with the `cluster-admin` role.

.Procedure

. Log in to the {product-title} CLI as a user with the `cluster-admin` role:
+
[source,terminal]
----
$ oc login -u kubeadmin https://<HOSTNAME>:6443
----

. Create a new project named `node-observability-operator`:
+
[source,terminal]
----
$ oc project node-observability-operator
----

. Create a custom resource file named `nodeobservability.yaml` that contains the following text:
+
[source,yaml]
----
    apiVersion: nodeobservability.olm.openshift.io/v1alpha1
    kind: NodeObservability
    metadata:
      name: cluster <1>
    spec:
      labels:
        "node-role.kubernetes.io/worker": ""

>       image: "brew.registry.redhat.io/rh-osbs/node-observability-agent:0.1.0-3"

Reviewer 2: `image` was removed from the spec starting https://github.com/openshift/node-observability-operator/pull/69 
But you're right, we need to update the readme in https://github.com/openshift/node-observability-operator/tree/main/docs#prepare-to-run-profiling-queries
In the meantime, you can use example from : https://github.com/openshift/node-observability-operator/blob/main/config/samples/nodeobservability_v1alpha1_nodeobservability-all.yaml

Author: Fixed.
